src/models/kernels.cu

```diff
@@ -1,71 +1,87 @@
-#include <cuda_runtime.h>
 #include <cuda_fp16.h>
+#include <cuda_runtime.h>
 #include <stdint.h>
 
 namespace Generators {
 namespace cuda {
 
-template <typename T>
-__global__ void UpdatePositionIds(T* positions, int batch_beam_size) {
-  int i = blockIdx.x * blockDim.x + threadIdx.x;
-  if (i < batch_beam_size)
-    positions[i]++;
+template <typename T> __global__ void UpdatePositionIds(T *positions, int batch_beam_size) {
+    int i = blockIdx.x * blockDim.x + threadIdx.x;
+    if (i < batch_beam_size)
+        positions[i]++;
 }
 
-template <typename T>
-void Launch_UpdatePositionIds(T* positions, int batch_beam_size, cudaStream_t stream) {
-  UpdatePositionIds<T><<<(batch_beam_size + 255) / 256, 256, 0, stream>>>(positions, batch_beam_size);
+template <typename T> void Launch_UpdatePositionIds(T *positions, int batch_beam_size, cudaStream_t stream) {
+    UpdatePositionIds<T><<<(batch_beam_size + 255) / 256, 256, 0, stream>>>(positions, batch_beam_size);
 }
 
-template void Launch_UpdatePositionIds(int32_t* positions, int batch_beam_size, cudaStream_t stream);
-template void Launch_UpdatePositionIds(int64_t* positions, int batch_beam_size, cudaStream_t stream);
+template void Launch_UpdatePositionIds(int32_t *positions, int batch_beam_size, cudaStream_t stream);
+template void Launch_UpdatePositionIds(int64_t *positions, int batch_beam_size, cudaStream_t stream);
 
 template <typename T>
-__global__ void UpdateAttentionMask(T* mask_data, const T* old_mask_data, int batch_beam_size, int current_length) {
-  int global_index = blockIdx.x * blockDim.x + threadIdx.x;
-  int i = global_index / current_length;
-  int j = global_index % current_length;
-  if (i < batch_beam_size) {
-    if (j < current_length - 1) {
-      mask_data[i * current_length + j] = old_mask_data[i * (current_length - 1) + j];
-    } else {
-      mask_data[i * current_length + j] = 1;
+__global__ void CopyAndUpdateAttentionMask(T *mask_data, const T *old_mask_data, int batch_beam_size,
+                                           int current_length, int max_length) {
+    int global_index = blockIdx.x * blockDim.x + threadIdx.x;
+    int i = global_index / current_length;
+    int j = global_index % current_length;
+    if (i < batch_beam_size) {
+        if (j < current_length - 1) {
+            mask_data[i * max_length + j] = old_mask_data[i * (current_length - 1) + j];
```

Comment:
>max_length [](http://example.com/codeflow?start=26&length=10)

is it max_length or current_length? Do you need a max_length for copy and update scenario?

Reply:
The new mask uses max length and the old mask uses current length because we allocate max length to mask and enable cuda graph in the second iteration